pyclam/manifold.py

```diff
@@ -410,42 +417,30 @@ def depth(self) -> int:
     def metric(self) -> Metric:
         return next(iter(self.clusters.keys())).metric
 
-    def _build_edges_matrix(self) -> None:
-        """ Calculates overlap for clusters in self in the naive way. """
-        logging.info(f'current depth: {self.depth}, {len(self.clusters)} clusters')
-        # TODO: Calculate memory cost of the distance matrix here.
-        clusters: List[Cluster] = list(self.clusters.keys())
-
-        argcenters = [c.argmedoid for c in clusters]
-        radii = np.asarray([c.radius for c in clusters], dtype=np.float64)
-
-        if len(clusters) <= BATCH_SIZE:
-            # TODO: Put this back, seeing weird errors.
-            distances = self.distance(
-                argcenters, 
-                argcenters,
-            )
-            differences = (distances.T - radii).T - radii
-            left, right = tuple(map(list, np.where(differences <= 0.)))
-
-            [clusters[l_].neighbors.update({clusters[r_]: distances[l_, r_]}) for l_, r_ in zip(left, right) if l_ != r_]
-        else:
-            for i in range(0, len(clusters), BATCH_SIZE):
-                batch_clusters = clusters[i: i + BATCH_SIZE]
-                batch_argcenters = [c.argmedoid for c in batch_clusters]
-                batch_radii = np.asarray([c.radius for c in batch_clusters], dtype=np.float64)
-
-                batch_distances = self.distance(batch_argcenters, argcenters)
-                batch_differences = ((batch_distances - radii).T - batch_radii).T
-
-                batch_left, batch_right = tuple(map(list, np.where(batch_differences <= 0.)))
-                [clusters[l_ + i].neighbors.update({clusters[r_]: batch_distances[l_, r_]}) for l_, r_ in zip(batch_left, batch_right) if (l_ + i) != r_]
-
-        return
-
     def build_edges(self) -> None:
```

Comment:
Is this the function that is only called at the bottom now? That should be specified if so

Reply:
This calls top-down, but I thought we discussed that the new function builds bottom-up, can you clarify?

Reply:
It gets called once from the bottom layer. This should indeed be specified.
It iterates over the graphs from the graph at depth 1 to itself. This is handled in in lines 437-442.

Interestingly, the helper _find_neighbors is constructed in such a way so as not to repeat any distance computations in layers over which the graph was already build so it does not even matter if Graph.build_edges gets called multiple times per layer. It will only do the incremental work necessary to complete the building of any graphs.